dhis-2/dhis-services/dhis-service-expression-parser/src/main/java/org/hisp/dhis/expressionparser/ExpressionVisitor.java

```diff
@@ -0,0 +1,596 @@
+    {
+        Double returnVal = null;
+
+        for ( ExprContext c : ctx.a1_n().expr() )
+        {
+            Double val = castDouble( visit( c ) );
+
+            if ( returnVal == null || val != null && ( val - returnVal ) * minmax > 0 )
+            {
+                returnVal = val;
+            }
+        }
+        return returnVal;
+    }
+
+    // -------------------------------------------------------------------------
+    // Other supportive functions
+    // -------------------------------------------------------------------------
+
+    /**
+     * Evaluates a one-argument double function, but returns null if the
+     * input value is null.
+     *
+     * @param ctx the parsing context with one expression.
+     * @param fn the function to evaluate.
+     * @return the function result, or null if input was null.
+     */
+    protected Double double1( ExprContext ctx, Function<Double, Double> fn )
+    {
+        Double d1 = castDouble( visit( ctx.expr( 0 ) ) );
+
+        return d1 == null
+            ? null
+            : fn.apply( d1 );
+    }
+
+    /**
+     * Evaluates a two-argument double function, but returns null if either of
+     * the input values is null.
+     *
+     * @param ctx the parsing context with two expressions.
+     * @param fn the function to evaluate.
+     * @return the function result, or null if either input was null.
+     */
+    protected Double double2( ExprContext ctx, BiFunction<Double, Double, Double> fn )
+    {
+        Double d1 = castDouble( visit( ctx.expr( 0 ) ) );
+        Double d2 = castDouble( visit( ctx.expr( 1 ) ) );
+
+        return double2( d1, d2, fn);
+    }
+
+    /**
+     * Evaluates a two-argument double function, but returns null if either of
+     * the input values is null.
+     *
+     * @param d1 the first function argument (could be null).
+     * @param d2 the second function argument (could be null).
+     * @param fn the function to evaluate.
+     * @return the function result, or null if either input was null.
+     */
+    protected Double double2( Double d1, Double d2, BiFunction<Double, Double, Double> fn )
+    {
+        return d1 == null || d2 == null
+            ? null
+            : fn.apply( d1, d2 );
+    }
+
+    /**
+     * Evaluates a two-argument boolean function, but returns null if either of
+     * the input values is null.
+     *
+     * @param ctx the parsing context with two expressions.
+     * @param fn the function to evaluate.
+     * @return the function result, or null if either input was null.
+     */
+    protected Boolean boolean1( ExprContext ctx, Function<Boolean, Boolean> fn )
+    {
+        Boolean b1 = castBoolean( visit( ctx.expr( 0 ) ) );
+
+        return b1 == null
+            ? null
+            : fn.apply( b1 );
+    }
+
+    /**
+     * Casts object as Integer, or throws an exception if we can't.
+     *
+     * @param object
+     * @return
+     */
+    protected Integer castInteger( Object object )
```

Comment:
For the moment, I've removed this method since it is unused. When supporting integers in the future (e.g., relative orgUnit levels) I may add this method back in, but I will do so in such a way that doesn't test doubles for strict equality.

Note that comparing doubles always works if the double value was an integer to begin with, or was formed by an equation only involving integers. It can fail only if the double value was built from fractions that cannot be represented in binary, for example 1.001 - 0.001. However it will always work if the fraction can be coded precisely in binary (within the number of bits in the fractional part of the double) such as 1/2, 1/4, 3/4, 1/8, 3/8, ..., for example 1.125 - 0.125. This is why I choose such fractions as JUnit test values for doubles -- especially if they will later be put in string values and compared, as they are in the test code for this service.

At any rate, I will avoid testing doubles for equality, to comply with current coding sensibilities.